File: studio/repositories/assets_repo.py
"""Project asset metadata persistence."""

import time

from studio import db
from studio.models import Asset, to_doc

_ALLOWED_TYPES = frozenset({"logo", "brand", "service", "team", "product"})


def allowed_asset_types():
    return sorted(_ALLOWED_TYPES)


def create_pending(
    project_id,
    *,
    asset_id,
    asset_type,
    s3_key,
    filename,
    content_type,
):
    if asset_type not in _ALLOWED_TYPES:
        raise ValueError("invalid asset type")
    now = time.time()
    row = Asset(
        assetId=asset_id,
        projectId=project_id,
        assetType=asset_type,
        filename=filename,
        contentType=content_type,
        s3Key=s3_key,
        status="pending",
        sizeBytes=None,
        width=None,
        height=None,
        uploadedAt=None,
        createdAt=now,
        updatedAt=now,
    )
    with db.session() as s:
        s.add(row)
        s.commit()
        return to_doc(row)


def get(project_id, asset_id):
    with db.session() as s:
        row = (
            s.query(Asset)
            .filter_by(projectId=project_id, assetId=asset_id)
            .first()
        )
        return to_doc(row)


def confirm(project_id, asset_id, *, size_bytes, width=None, height=None):
    now = time.time()
    with db.session() as s:
        s.query(Asset).filter_by(projectId=project_id, assetId=asset_id).update(
            {
                "status": "ready",
                "sizeBytes": size_bytes,
                "width": width,
                "height": height,
                "uploadedAt": now,
                "updatedAt": now,
            }
        )
        s.commit()
    return get(project_id, asset_id)


def list_for_project(project_id):
    with db.session() as s:
        rows = (
            s.query(Asset)
            .filter_by(projectId=project_id, status="ready")
            .order_by(Asset.uploadedAt.desc())
            .all()
        )
        return [to_doc(r) for r in rows]

File: src/server/repos/assetsRepo.ts
import { prisma } from "../db";
import { toDoc } from "./doc";

const ALLOWED_TYPES = new Set(["logo", "brand", "service", "team", "product"]);

function encodeAssetType(assetType: string, slotKey?: string | null): string {
  if (!slotKey) return assetType;
  return `${assetType}@${slotKey}`;
}

export function decodeSlotKey(asset: Record<string, unknown>): string | null {
  const raw = String(asset.assetType ?? "");
  const idx = raw.indexOf("@");
  return idx >= 0 ? raw.slice(idx + 1) : null;
}

export function allowedAssetTypes(): string[] {
  return [...ALLOWED_TYPES].sort();
}

export async function createPending(
  projectId: string,
  data: {
    assetId: string;
    assetType: string;
    s3Key: string;
    filename: string;
    contentType: string;
    slotKey?: string | null;
  },
): Promise<Record<string, unknown>> {
  if (!ALLOWED_TYPES.has(data.assetType)) throw new Error("invalid asset type");
  const now = Date.now() / 1000;
  const row = await prisma().projectAsset.create({
    data: {
      assetId: data.assetId,
      projectId,
      assetType: encodeAssetType(data.assetType, data.slotKey),
      filename: data.filename,
      contentType: data.contentType,
      s3Key: data.s3Key,
      status: "pending",
      createdAt: now,
      updatedAt: now,
    },
  });
  return toDoc(row) as Record<string, unknown>;
}

export async function get(
  projectId: string,
  assetId: string,
): Promise<Record<string, unknown> | null> {
  const row = await prisma().projectAsset.findFirst({ where: { projectId, assetId } });
  return toDoc(row) as Record<string, unknown> | null;
}

export async function confirm(
  projectId: string,
  assetId: string,
  data: { sizeBytes: number; width?: number | null; height?: number | null },
): Promise<Record<string, unknown> | null> {
  const now = Date.now() / 1000;
  await prisma().projectAsset.updateMany({
    where: { projectId, assetId },
    data: {
      status: "ready",
      sizeBytes: data.sizeBytes,
      width: data.width ?? null,
      height: data.height ?? null,
      uploadedAt: now,
      updatedAt: now,
    },
  });
  return get(projectId, assetId);
}

export async function listForProject(projectId: string): Promise<Record<string, unknown>[]> {
  const rows = await prisma().projectAsset.findMany({
    where: { projectId, status: "ready" },
    orderBy: { uploadedAt: "desc" },
  });
  return rows.map((row) => toDoc(row) as Record<string, unknown>);
}
